lib/Dialect/ESI/cosim/cosim_dpi_server/DpiEntryPoints.cpp

```diff
@@ -0,0 +1,213 @@
+  if (numElems * expectedElemSize != totalBytes) {
+    printf("DPI-C: ERROR: passed array argument that doesn't have expected "
+           "element-size: expected=%d actual=%d\n",
+           expectedElemSize, elemSize);
+    return -4;
+  }
+  return 0;
+}
+
+// ---- DPI entry points ----
+
+// Register simulated device endpoints.
+// - return 0 on success, non-zero on failure (duplicate EP registered).
+DPI int sv2cCosimserverEpRegister(int endpointId, long long sendTypeId,
+                                  int sendTypeSize, long long recvTypeId,
+                                  int recvTypeSize) {
+  // Ensure the server has been constructed.
+  sv2cCosimserverInit();
+  try {
+    // Then register with it.
+    server->endpoints.registerEndpoint(endpointId, sendTypeId, sendTypeSize,
+                                       recvTypeId, recvTypeSize);
+    return 0;
+  } catch (const std::runtime_error &rt) {
+    fprintf(stderr, "%s\n", rt.what());
+    return -1;
+  }
+}
+
+// Attempt to recieve data from a client.
+//   - Returns negative when call failed (e.g. EP not registered).
+//   - If no message, return 0 with dataSize == 0.
+//   - Assumes buffer is large enough to contain entire message. Fails if not
+//     large enough. (In the future, will add support for getting the message
+//     into a fixed-size buffer over multiple calls.)
+DPI int sv2cCosimserverEpTryGet(unsigned int endpointId,
+                                // NOLINTNEXTLINE(misc-misplaced-const)
+                                const svOpenArrayHandle data,
+                                unsigned int *dataSize) {
+  if (server == nullptr)
+    return -1;
+
+  Endpoint *ep = server->endpoints[endpointId];
+  if (!ep) {
+    fprintf(stderr, "Endpoint not found in registry!\n");
+    return -4;
+  }
+
+  Endpoint::BlobPtr msg;
+  // Poll for a message.
+  if (ep->getMessageToSim(msg)) {
+
+    // Do the validation only if there's a message available. Since the
+    // simulator is going to poll up to every tick and there's not going to be
+    // a message most of the time, this is important for performance.
+
+    if (validateSvOpenArray(data, sizeof(int8_t)) != 0) {
+      printf("ERROR: DPI-func=%s line=%d event=invalid-sv-array\n", __func__,
+             __LINE__);
+      return -2;
+    }
+
+    // Detect or verify size of buffer.
+    if (*dataSize == ~0u) {
+      *dataSize = svSizeOfArray(data);
+    } else if (*dataSize > (unsigned)svSizeOfArray(data)) {
+      printf("ERROR: DPI-func=%s line %d event=invalid-size (max %d)\n",
+             __func__, __LINE__, (unsigned)svSizeOfArray(data));
+      return -3;
+    }
+    // Verify it'll fit.
+    if (msg->size() > *dataSize) {
+      printf("ERROR: Message size too big to fit in RTL buffer\n");
+      return -5;
+    }
+
+    // Copy the message data.
+    size_t i;
+    size_t msgSize = msg->size();
+    for (i = 0; i < msgSize; ++i) {
+      auto b = msg->at(i);
```

Comment:
Please don't use at(), it implies exceptions.

Reply:
did not know that!